### ISSUE TITLE
The reason why we get unalign outputs from vllm and hf?

### ISSUE THREAD

**Issue by AmazeQiu** [vllm-project/vllm#1334]
Hey guys! Have you ever noticed that the outputs from vllm and hf is unalign even if you use the greedy search strategy?

I tried to run the test scripts with pytest in vllm project:
[https://github.com/vllm-project/vllm/blob/main/tests/kernels/test_attention.py](url)
[https://github.com/vllm-project/vllm/blob/main/tests/kernels/test_layernorm.py](url)
yes, they were all passed.
then I found the assert expression is like this:
`assert torch.allclose(output, ref_output, atol=1e-3, rtol=1e-5)`
`assert torch.allclose(out, ref_out, atol=1e-2, rtol=1e-5)`

Then I changed the atol to 1e-4 and 1e-3, the tests failed!

So I wonder, is it normal to have such a big numerical difference between vllm cuda operators and torch operators?

That's why when I used baichuan-13B to run the model test script, it failed and I found that the front part of the two generated sentences are the same but diverged in the middle.  
I guess it's because the numerical difference will become bigger and bigger with the inference steps?
So is there any way to solve this problem?

**Comment by casper-hansen** [comment]
This is not a problem, and we should not see it as being incorrect. A Huggingface model does not constitute a correct model, neither does a vLLM model. You are correct in that there are slight differences in logits which leads to slightly different outputs and outcomes.

You should instead measure perplexity, MMLU, and other benchmarks to evaluate if the models drop in performance.

**Comment by AmazeQiu** [comment]
> This is not a problem, and we should not see it as being incorrect. A Huggingface model does not constitute a correct model, neither does a vLLM model. You are correct in that there are slight differences in logits which leads to slightly different outputs and outcomes.
> 
> You should instead measure perplexity, MMLU, and other benchmarks to evaluate if the models drop in performance.

Thank you for reply! You're right , difference dosen't mean it's incorrect. 
In our situtation, we just use the beam search or greedy search strategy and set temperature to 0.  We have our own test set to evaluate the performance. We use hf and peft to finetune the model and plan to depoly it with vllm. Now it's just  the difference of the performance that conserns us.

**Comment by kar9999** [comment]
I also have the same question. After fine-tuning the Baichuan model, we used VLLM to accelerate, and I don't know why the quality of our answers decreased compared to that of the Huggingface.

**Comment by AmazeQiu** [comment]
> I also have the same question. After fine-tuning the Baichuan model, we used VLLM to accelerate, and I don't know why the quality of our answers decreased compared to that of the Huggingface.

For the latest version of vllm, now the performance of the two is really close. Now we use vllm to accelerate, it works well for us.

**Comment by hmellor** [comment]
Closing this issue as stale as there has been no discussion in the past 3 months.

If you are still experiencing the issue you describe, feel free to re-open this issue.
